enhancements/oc/image-content-source-policy-awareness.md

```diff
@@ -0,0 +1,109 @@
+        So here, we'd need a different `--cluster-icsp` boolean and `--icsp-file` string flag.  The flags are adding up here, and that is not desireable. 
+        * string `--image-content-source aregistry/arepo/arelease`. I don't like this, because it would be redundant for a user to run something like this: 
+        `oc adm release extract --command oc myreg:5000/myrepo/release:tag --image-content-source myreg:5000/myrepo/release`.  
+        * Instead, a boolean `--set-prefix` would allow a user to specify "I want to use the prefix of the release image I have specified, 
+        rather than any underlying image reference."
+* This proposal is for the `FLAGS`, that `oc` will try ICSP, user-given image, then fall back on underlying image-reference from user-given image. 
+* See `User Stories` below for examples. 
+
+## Summary
+
+ICSP allows OpenShift (CVO, CRI-O) to check down a list of possible mirrors to find an image with the matching digest it is
+looking for.  `oc` should do the same.  If no ICSP found, then try user-given image.  If that doesn't succeed, then use underlying
+image-reference from user-given image.
+
+## Proposal
+
+* `oc adm release mirror` writes an ICSP file to the current directory or wherever you specify 
+* Add logic to `oc adm release` to become aware of ICSP in cluster
+* Add logic to `oc adm release` to use the ICSP to complete extracts, mirroring, must-gather
+
+## User Stories
+
+A user runs:
+1. `oc adm must-gather`
+    * `oc` will look for ICSP from cluster (no flag necessary here, since this cmd is always run while connected to a cluster) and will use the mirror from ICSP to launch the must-gather pod.  If no ICSP found, will look in `-n openshift` for must-gather imagestream.
+2. `oc adm release extract --tools registry.example.com/repo/release:tag`
+    * `oc` will look for ICSP and if found, will extract from the ICSP mirror (aregistry/arepo/release:tag-toolsha) rather than from the image reference that the user will not
+    have access to if in disconnected environment.  If ICSP not found, will extract from the user-given registry.example.com/repo/tool:digest.  
+    If neither of these succeed, will proceed to use the image-reference from the user-given image.  The extract will succeed if user has access and permission to the
+    underlying image registry/repo/name.
+2. `oc adm release extract --icsp-file /path/to/icsp.yaml --tools anightly/release:tag`
+    * `oc` will try to use data from the icsp file, it will extract from the ICSP mirror (aregistry/arepo/release:tag-toolsha) rather than from the image reference.
+    It will not try other sources, if the --icsp-file flag is provided.  
+3. `oc adm release mirror --icsp-to /path/to/file aregistry/arepo/release:tag --to anotherreg/arepo/release`
+    * `oc` will write an ICSP file to the given path.  If no --icsp-to flag, `oc` will write the ICSP file to the current directory. 
```

Comment:
Current directory or if `--to-dir` is specified write there.

Reply:
updated